test/modeles/risques.spec.js

```diff
@@ -120,8 +120,38 @@ describe('Les risques liés à une homologation', () => {
       });
       const risques = new Risques({ risquesGeneraux: [{ id: 'unRisque', niveauGravite: 'grave' }] }, referentiel);
       risques.risquesGeneraux.parNiveauGravite = () => ({ grave: [{ description: 'Un risque' }] });
+      risques.risquesSpecifiques.parNiveauGravite = (risquesParNiveau) => risquesParNiveau;
 
       expect(risques.parNiveauGravite()).to.eql({ grave: [{ description: 'Un risque' }] });
     });
+
+    it('ajoutent les risques spécifiques triés', () => {
+      const referentiel = Referentiel.creeReferentiel({
+        niveauxGravite: { grave: {} },
+      });
+      const risques = new Risques({ risquesSpecifiques: [{ description: 'Un risque', niveauGravite: 'grave' }] }, referentiel);
+      risques.risquesGeneraux.parNiveauGravite = () => ({});
+      risques.risquesSpecifiques.parNiveauGravite = () => ({ grave: [{ description: 'Un risque' }] });
+
+      expect(risques.parNiveauGravite()).to.eql({ grave: [{ description: 'Un risque' }] });
+    });
```

Comment:
J'ai été gêné par ce test, car l'assertion est identique à celle du test au-dessus.

Plus généralement dans ce `describe('sur une demande des risques par niveau de gravité'` je suis aussi gêné par les mocks sur `risques.risquesGeneraux.parNiveauGravite` et `risques.risquesSpecifiques.parNiveauGravite`.
Et j'ai bien compris que ces mocks étaient présents dès le premier test `it('délègue la demande aux risques généraux'` qui ne vient pas de cette PR.

Mais on est dans un périmètre restreint (la grappe des risques) et j'aimerais exercer toute la grappe dans les tests.

Je propose de changer le test en-dessous de celui-ci par une version sans mock :

```js
    it('fusionnent les risque généraux et spécifiques (version sans mock)', () => {
      const referentiel = Referentiel.creeReferentiel({
        risques: { unRisque: { description: 'Un risque' } },
        niveauxGravite: { grave: {} },
      });

      const risques = new Risques({
        risquesGeneraux: [{ id: 'unRisque', niveauGravite: 'grave' }],
        risquesSpecifiques: [{ description: 'Un risque deux', niveauGravite: 'grave' }],
      }, referentiel);

      expect(risques.parNiveauGravite()).to.eql({ grave: [
        { id: 'unRisque', niveauGravite: 'grave', description: 'Un risque' },
        { niveauGravite: 'grave', description: 'Un risque deux' },
      ] });
    });

```

Et de supprimer le test `it('ajoutent les risques spécifiques triés'`.

Je pense même que j'aurais envie de supprimer le test `it('délègue la demande aux risques généraux`.

En résumé : dans ce `describe`, ne garder qu'un seul test, et qu'il soit sans mock.

Reply:
L'utilisation de mock est présente pour montrer la délégation, le vrai comportement est testé ailleurs.
Je préfère laisser les 3 tests, je les ai renommés et j'ai créé une fonction pour bouchonner. Je te laisse voir si cela convient.

Reply:
Merci pour la mise à jour. Désolé, pour moi ça ne change rien.

Je comprends qu'on dégaine les mocks pour le cas des tests d'`Homologation` quand il s'agit de ne pas avoir une phase « Arrange » du test qui fait 10 lignes et qui positionne des variables très éloignées de l'`Homologation` testée.

Mais là… la version du test sans mock est plus courte que la version avec mock.
Et pour moi, elle est aussi plus simple à comprendre.

Quand je regarde les tests qui restent dans le describe :
```js
describe('sur une demande des risques par niveau de gravité')
  ils('récupèrent ceux des risques généraux')
  ils('récupèrent ceux des risques spécifiques')
  ils('fusionnent les risques récupérés généraux et spécifiques')
```

Je trouve que le 3eme test suffit sur un comportement si simple.

Et je remets les 2 versions avec et sans mock :

```js
    it('fusionnent les risque généraux et spécifiques (version sans mock)', () => {
      const referentiel = Referentiel.creeReferentiel({
        risques: { unRisque: { description: 'Un risque' } },
        niveauxGravite: { grave: {} },
      });

      const risques = new Risques({
        risquesGeneraux: [{ id: 'unRisque', niveauGravite: 'grave' }],
        risquesSpecifiques: [{ description: 'Un risque deux', niveauGravite: 'grave' }],
      }, referentiel);

      expect(risques.parNiveauGravite()).to.eql({ grave: [
        { id: 'unRisque', niveauGravite: 'grave', description: 'Un risque' },
        { niveauGravite: 'grave', description: 'Un risque deux' },
      ] });
    });
```

```js
    ils('fusionnent les risques récupérés généraux et spécifiques', () => {
      const referentiel = Referentiel.creeReferentiel({
        risques: { unRisque: { description: 'Un risque' } },
        niveauxGravite: { grave: {} },
      });
      const risques = new Risques({
        risquesGeneraux: [{ id: 'unRisque', niveauGravite: 'grave' }],
        risquesSpecifiques: [{ description: 'Un risque deux', niveauGravite: 'grave' }],
      }, referentiel);
      risques.risquesGeneraux.parNiveauGravite = () => ({ grave: [{ description: 'Un risque' }] });
      risques.risquesSpecifiques.parNiveauGravite = (risquesParNiveauGravite) => {
        expect(risquesParNiveauGravite).to.eql({ grave: [{ description: 'Un risque' }] });
        return { grave: [{ description: 'Un risque' }, { description: 'Un risque deux' }] };
      };

      expect(risques.parNiveauGravite()).to.eql({ grave: [{ description: 'Un risque' }, { description: 'Un risque deux' }] });
    });
```

Je suggère qu'on continue à l'oral si tu penses que nous devons encore discuter 😸

Reply:
@cadichris – ça m'a semblé important par le passé qu'on installe des mocks pour alléger des tests qui traversaient plusieurs couches. On a adopté une approche radicale de ne tester qu'une couche à la fois, d'où quelques tests de délégations qui se retrouvent dans le code.

Maintenant qu'on est à peu près au clair sur ce standard, on pourra recommencer à introduire de la nuance et notamment à regarder un peu plus finement le compromis entre couplage (un test qui a besoin de plein d'objets pour initialiser le contexte) et fragilité (un test qui peut continuer à passer sans détecter un changement d'API).

En attendant, ça me va bien qu'on laisse comme Jacques a dit ;)

Reply:
> Quand je regarde les tests qui restent dans le describe :
>
> ```
> describe('sur une demande des risques par niveau de gravité')
>  ils('récupèrent ceux des risques généraux')
>  ils('récupèrent ceux des risques spécifiques')
>  ils('fusionnent les risques récupérés généraux et spécifiques')
>```
>
> Je trouve que le 3eme test suffit sur un comportement si simple.

Je suis d'accord on peut en laisser 1 : le dernier.
Je préfère quand même la version mocker car ce que cela teste c'est la récupération et la fusion et je ne souhaite pas que ça fail quand dans risque on ajout une nouvelle propriété par exemple

Reply:
Faisons comme ça, ça me va bien 👍